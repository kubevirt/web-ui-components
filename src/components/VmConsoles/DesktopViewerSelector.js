import React from 'react';
import PropTypes from 'prop-types';
import { DesktopViewer } from '@patternfly/react-console';
import { Alert } from 'patternfly-react';
import { get } from 'lodash';

import { FormFactory } from '../Form';
import { settingsValue } from '../../k8s/selectors';
import { DEFAULT_RDP_PORT } from '../../constants';
import { RdpServiceNotConfigured } from './VmConsoles';
import {
  SELECT_NETWORK_INTERFACE,
  NIC,
  GUEST_AGENT_WARNING,
  NO_IP_ADDRESS,
  RDP_NETWORK_INTERFACE_HELP,
} from './strings';
import { NIC_KEY } from './constants';
import { NETWORK_TYPE_MULTUS, NETWORK_TYPE_POD } from '../Wizard/CreateVmWizard/constants';
import { getNetworks } from '../../k8s/vmBuilder';

const getForm = networks => ({
  [NIC_KEY]: {
    id: 'nic-dropdown',
    title: NIC,
    type: 'dropdown',
    defaultValue: SELECT_NETWORK_INTERFACE,
    choices: networks,
    help: RDP_NETWORK_INTERFACE_HELP,
  },
});

const getVmRdpNetworks = (vm, vmi) => {
  const networks = getNetworks(vm).filter(n => n.multus || n.pod);
  return get(vmi, 'status.interfaces', [])
    .filter(i => networks.some(n => n.name === i.name))
    .map(i => {
      let ip = i.ipAddress;
      if (i.ipAddress) {
        const subnetIndex = i.ipAddress.indexOf('/');
        if (subnetIndex > 0) {
          ip = i.ipAddress.slice(0, subnetIndex);
        }
      }
      const network = networks.find(n => n.name === i.name);
      return {
        name: i.name,
        type: get(network, 'multus') ? NETWORK_TYPE_MULTUS : NETWORK_TYPE_POD,
        ip,
      };
    });
};

const getDefaultNetwork = networks => {
  if (networks.length === 1) {
    return networks[0];
  }
  if (networks.length > 1) {
    return (
      networks.find(n => n.type === NETWORK_TYPE_MULTUS && n.ipAddress) ||
      networks.find(n => n.type === NETWORK_TYPE_MULTUS)
    );
  }
  return null;
};

export class DesktopViewerSelector extends React.Component {
  constructor(props) {
    super(props);
    const networks = getVmRdpNetworks(props.vm, props.vmi);
    this.state = {
      [NIC_KEY]: {
        value: getDefaultNetwork(networks),
      },
    };
  }

  onFormChange = (newValue, target) => {
    this.setState({
      [target]: newValue,
    });
  };

  render() {
    const networks = getVmRdpNetworks(this.props.vm, this.props.vmi);
    let content;
    switch (get(settingsValue(this.state, NIC_KEY), 'type')) {
      case NETWORK_TYPE_MULTUS:
        if (!this.props.guestAgent) {
          content = <Alert type="warning">{GUEST_AGENT_WARNING}</Alert>;
        } else if (!settingsValue(this.state, NIC_KEY).ip) {
          content = <Alert type="warning">{`${NO_IP_ADDRESS} ${settingsValue(this.state, NIC_KEY).name}`}</Alert>;
        } else {
          const rdp = {
            address: settingsValue(this.state, NIC_KEY).ip,
            port: DEFAULT_RDP_PORT,
          };
          content = <DesktopViewer rdp={rdp} />;
        }
        break;
      case NETWORK_TYPE_POD:
        content =
          this.props.rdpServiceManual || this.props.vncServiceManual ? (
            <DesktopViewer rdp={this.props.rdpServiceManual} vnc={this.props.vncServiceManual} />
          ) : (
            <RdpServiceNotConfigured vm={this.props.vm} />
          );
        break;
      default:
        // eslint-disable-next-line no-console
        console.warn(`Unknown network interface type ${get(settingsValue(this.state, NIC_KEY), 'type')}`);
    }

    return (
      <div className="kubevirt-desktop-viewer-selector">
        <FormFactory fields={getForm(networks)} fieldsValues={this.state} onFormChange={this.onFormChange} />
        {content}
      </div>
    );
  }
}

DesktopViewerSelector.defaultProps = {
  guestAgent: false,
  rdpServiceManual: null,
  vncServiceManual: null,
};

DesktopViewerSelector.propTypes = {
  vm: PropTypes.object.isRequired,
  vmi: PropTypes.object.isRequired,
  guestAgent: PropTypes.bool,
  rdpServiceManual: PropTypes.object,
  vncServiceManual: PropTypes.object,
};
