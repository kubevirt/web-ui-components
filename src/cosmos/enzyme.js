import createContext from 'react-cosmos-test/enzyme';

import paths from '../../config/paths';

export default function createTestContext(args) {
  args.proxies = args.proxies || []; // workaround of react-cosmos-test bug
  return createContext({
    ...args,
    cosmosConfigPath: `${paths.config}/cosmos.config.js`,
  });
}
