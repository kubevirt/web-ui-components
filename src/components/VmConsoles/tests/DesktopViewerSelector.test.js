import React from 'react';
import { shallow } from 'enzyme';

import { DesktopViewerSelector } from '../DesktopViewerSelector';
import { default as desktopViewerSelectorFixture } from '../fixtures/DesktopViewerSelector.fixture';

describe('<DesktopViewerSelector />', () => {
  it('renders correctly', () => {
    const component = shallow(<DesktopViewerSelector {...desktopViewerSelectorFixture[0].props} />);
    expect(component).toMatchSnapshot();
  });
});
