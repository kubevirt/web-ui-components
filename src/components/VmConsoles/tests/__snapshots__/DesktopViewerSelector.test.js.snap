// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DesktopViewerSelector /> renders correctly 1`] = `
<div
  className="kubevirt-desktop-viewer-selector"
>
  <FormFactory
    controlSize={5}
    fields={
      Object {
        "nic": Object {
          "choices": Array [
            Object {
              "ip": "192.168.1.0",
              "name": "eth0",
              "type": "pod",
            },
          ],
          "defaultValue": "--- Select Network Interface ---",
          "help": "The network interface to be used for accessing the RDP console",
          "id": "nic-dropdown",
          "title": "Network Interface",
          "type": "dropdown",
        },
      }
    }
    fieldsValues={
      Object {
        "nic": Object {
          "value": Object {
            "ip": "192.168.1.0",
            "name": "eth0",
            "type": "pod",
          },
        },
      }
    }
    horizontal={true}
    labelSize={3}
    onFormChange={[Function]}
    showLabels={true}
    textPosition="text-right"
  />
  <RdpServiceNotConfigured
    vm={
      Object {
        "apiVersion": "kubevirt.io/v1alpha3",
        "kind": "VirtualMachine",
        "metadata": Object {
          "annotations": Object {
            "description": "Lorem ipsum dolor sit amet, consectetur adipiscing elit. Donec lacus nibh, convallis vel nunc id,tempus vulputate augue. Proin eget nisl vel ante tincidunt accumsan vel at elit. Fusce eget tincidunt sem. Fusce cursus orci vitae nisl hendrerit mollis. Nullam at nulla ut ipsum malesuada laoreet a sit amet est.",
          },
          "clusterName": "",
          "creationTimestamp": "2018-11-06T14:32:07Z",
          "generation": 1,
          "labels": Object {
            "flavor.template.kubevirt.io/small": "true",
            "os.template.kubevirt.io/fedora29": "true",
            "vm.kubevirt.io/template": "fedora-generic",
            "vm.kubevirt.io/template-namespace": "default",
            "workload.template.kubevirt.io/generic": "true",
          },
          "name": "full-vm",
          "namespace": "default",
          "resourceVersion": "10390764",
          "selfLink": "/apis/kubevirt.io/v1alpha3/namespaces/default/virtualmachines/cloudinit-test-vm",
          "uid": "bcc1d0b1-e1d0-11e8-82b4-54ee7586b9c3",
        },
        "spec": Object {
          "dataVolumeTemplates": Array [
            Object {
              "metadata": Object {
                "name": "dv-template",
              },
              "spec": Object {
                "pvc": Object {
                  "accessModes": Array [
                    "ReadWriteOnce",
                  ],
                  "resources": Object {
                    "requests": Object {
                      "storage": "1G",
                    },
                  },
                },
                "source": Object {
                  "pvc": Object {
                    "name": "fooname",
                    "namespace": "foonamespace",
                  },
                },
              },
            },
          ],
          "running": true,
          "template": Object {
            "spec": Object {
              "domain": Object {
                "cpu": Object {
                  "cores": 2,
                },
                "devices": Object {
                  "disks": Array [
                    Object {
                      "disk": Object {
                        "bus": "virtio",
                      },
                      "name": "rootdisk",
                    },
                    Object {
                      "disk": Object {
                        "bus": "virtio",
                      },
                      "name": "cloudinitdisk",
                    },
                    Object {
                      "disk": Object {
                        "bus": "virtio",
                      },
                      "name": "datavolumedisk",
                    },
                    Object {
                      "disk": Object {
                        "bus": "virtio",
                      },
                      "name": "datavolumetemplatedisk",
                    },
                    Object {
                      "disk": Object {
                        "bus": "virtio",
                      },
                      "name": "pvcdisk",
                    },
                  ],
                  "interfaces": Array [
                    Object {
                      "bridge": Object {},
                      "name": "eth0",
                    },
                  ],
                  "rng": Object {},
                },
                "resources": Object {
                  "requests": Object {
                    "memory": "2G",
                  },
                },
              },
              "networks": Array [
                Object {
                  "name": "eth0",
                  "pod": Object {},
                },
              ],
              "terminationGracePeriodSeconds": 0,
              "volumes": Array [
                Object {
                  "containerDisk": Object {
                    "image": "kubevirt/cirros-registry-disk-demo",
                  },
                  "name": "rootdisk",
                },
                Object {
                  "dataVolume": Object {
                    "name": "datavolume-url",
                  },
                  "name": "datavolumedisk",
                },
                Object {
                  "dataVolume": Object {
                    "name": "dv-template",
                  },
                  "name": "datavolumetemplatedisk",
                },
                Object {
                  "name": "pvcdisk",
                  "persistentVolumeClaim": Object {
                    "claimName": "disk-one",
                  },
                },
                Object {
                  "cloudInitNoCloud": Object {
                    "userData": "#cloud-config
users:
  - name: root
    ssh-authorized-keys: |-
      AAAAB3NzaC1yc2EAAAABIwAAAQEAklOUpkDHrfHY17SbrmTIpNLTGK9Tjom/BWDSU
      GPl+nafzlHDTYW7hdI4yZ5ew18JH4JW9jbhUFrviQzM7xlELEVf4h9lFX5QVkbPppSwg0cda3
      Pbv7kOdJ/MTyBlWXFCR+HAo3FXRitBqxiX1nKhXpHAZsMciLq8V6RjsNAQwdsdMFvSlVK/7XA
      t3FaoJoAsncM1Q9x5+3V0Ww68/eIFmb1zuUFljQJKprrX88XypNDvjYNby6vw/Pb0rwert/En
      mZ+AW4OZPnTPI89ZPmVMLuayrD2cE86Z/il8b+gw3r3+1nKatmIkjn2so1d01QraTlMqVSsbx
      NrRFi9wrf+M7Q==
hostname: cloudinit-test
",
                  },
                  "name": "cloudinitdisk",
                },
              ],
            },
          },
        },
        "status": Object {
          "created": true,
          "ready": true,
        },
      }
    }
  />
</div>
`;
